### ISSUE TITLE
Unable to debug some web assembly apps from visual studio

### ISSUE THREAD

**Issue by musictopia2** [dotnet/aspnetcore#22628]
There is a web assembly app that I am unable to debug.
For example, I set a breakpoint for the counter and no matter how many times i click it, it never steps into.  The code still runs.  The problem is so bad i cannot even debug from the browser tools because the files don't even appear.  tried to follow the directions for chrome to have the files appear but they still don't appear.  I am using 2 external class libraries.  I would be disappointed if somehow using external class libraries cause the breakpoints to never get hit.

**Comment by mkArtakMSFT** [comment]
Thank you for filing this issue. In order for us to investigate this issue, please provide a minimalistic repro project (ideally a GitHub repo) that illustrates the problem.

**Comment by imtrobin** [comment]
I'm seen similar issue, this is repro. set breakpoint at Index.razor.cs , the OnAfterRenderAsync is not hit

https://filedn.com/l3TGy7Y83c247u0RDYa9fkp/temp/code/TestBlazorLifeCycle.zip

**Comment by imtrobin** [comment]
```
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://127.0.0.1:39546
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: D:\Data\Desktop\TestBlazorLifeCycle\TestBlazorLifeCycle
dbug: Microsoft.Extensions.Hosting.Internal.Host[1]
      Hosting starting
dbug: Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer[2]
      Failed to locate the development https certificate at '(null)'.
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://127.0.0.1:50142
dbug: Microsoft.AspNetCore.Hosting.Diagnostics[0]
      Loaded hosting startup assembly Microsoft.AspNetCore.Components.WebAssembly.DebugProxy
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\User\.nuget\packages\microsoft.aspnetcore.components.webassembly.devserver\3.2.0\tools\BlazorDebugProxy\
dbug: Microsoft.Extensions.Hosting.Internal.Host[2]
      Hosting started
dbug: Microsoft.AspNetCore.Server.Kestrel[39]
      Connection id "0HM0DITMUO9NU" accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel[1]
      Connection id "0HM0DITMUO9NU" started.
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://127.0.0.1:50142/ws-proxy?browser=ws%3A%2F%2Flocalhost%3A58025%2Fdevtools%2Fbrowser%2F499dcef4-6e37-44c8-81dd-4d121a15b807  
dbug: Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware[0]
      Wildcard detected, all requests with hosts will be allowed.
dbug: Microsoft.AspNetCore.Routing.Matching.DfaMatcher[1001]
      1 candidate(s) found for the request path '/ws-proxy'
dbug: Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware[1]
      Request matched endpoint '/ws-proxy HTTP: GET'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint '/ws-proxy HTTP: GET'
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      DevToolsProxy: Starting on ws://localhost:58025/devtools/browser/499dcef4-6e37-44c8-81dd-4d121a15b807
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      DevToolsProxy: IDE waiting for connection on ws://localhost:58025/devtools/browser/499dcef4-6e37-44c8-81dd-4d121a15b807
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      DevToolsProxy: Client connected on ws://localhost:58025/devtools/browser/499dcef4-6e37-44c8-81dd-4d121a15b807
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      Default context created, clearing state and sending events
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      Default context created, clearing state and sending events
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      proxying Debugger.scriptParsed (A6E1284E44129E9D0FDBE2D940ECD1C3) https://localhost:44355/_framework/blazor.webassembly.js {
        "scriptId": "8",
        "url": "https://localhost:44355/_framework/blazor.webassembly.js",
        "startLine": 0,
        "startColumn": 0,
        "endLine": 0,
        "endColumn": 59987,
        "executionContextId": 2,
        "hash": "020fa35e6399e95f6e67c7d33365db28091c6bb8",
        "executionContextAuxData": {
          "isDefault": true,
          "type": "default",
          "frameId": "E656A2E8650FC2EC12B80ACE7D527809"
        },
        "isLiveEdit": false,
        "sourceMapURL": "",
        "hasSourceURL": false,
        "isModule": false,
        "length": 59987,
        "scriptLanguage": "JavaScript"
      }
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      proxying Debugger.scriptParsed (A6E1284E44129E9D0FDBE2D940ECD1C3) https://localhost:44355/_framework/wasm/dotnet.3.2.0.js {
        "scriptId": "10",
        "url": "https://localhost:44355/_framework/wasm/dotnet.3.2.0.js",
        "startLine": 0,
        "startColumn": 0,
        "endLine": 280,
        "endColumn": 0,
        "executionContextId": 2,
        "hash": "24cdd081aace7a5f7d0f3151a5ae836655a7b7b6",
        "executionContextAuxData": {
          "isDefault": true,
          "type": "default",
          "frameId": "E656A2E8650FC2EC12B80ACE7D527809"
        },
        "isLiveEdit": false,
        "sourceMapURL": "",
        "hasSourceURL": false,
        "isModule": false,
        "length": 222435,
        "scriptLanguage": "JavaScript"
      }
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      proxying Debugger.scriptParsed (A6E1284E44129E9D0FDBE2D940ECD1C3) https://localhost:44355/_framework/wasm/dotnet.wasm {
        "scriptId": "11",
        "url": "https://localhost:44355/_framework/wasm/dotnet.wasm",
        "startLine": 0,
        "startColumn": 0,
        "endLine": 0,
        "endColumn": 0,
        "executionContextId": 2,
        "hash": "3fb75160ab1f4e4e82675bc4cd924d3481abe278",
        "executionContextAuxData": {
          "isDefault": true,
          "type": "default",
          "frameId": "E656A2E8650FC2EC12B80ACE7D527809"
        },
        "isLiveEdit": false,
        "sourceMapURL": "",
        "hasSourceURL": true,
        "isModule": false,
        "length": 0,
        "codeOffset": 12171,
        "scriptLanguage": "WebAssembly"
      }
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      	sending file:///D:/Data/Desktop/TestBlazorLifeCycle/TestBlazorLifeCycle/Program.cs 2 A6E1284E44129E9D0FDBE2D940ECD1C3
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      	sending file:///D:/Data/Desktop/TestBlazorLifeCycle/TestBlazorLifeCycle/Shared/NavMenu.razor 2 A6E1284E44129E9D0FDBE2D940ECD1C3
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      	sending file:///D:/Data/Desktop/TestBlazorLifeCycle/TestBlazorLifeCycle/Shared/SurveyPrompt.razor 2 A6E1284E44129E9D0FDBE2D940ECD1C3
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      	sending file:///D:/Data/Desktop/TestBlazorLifeCycle/TestBlazorLifeCycle/Pages/Counter.razor 2 A6E1284E44129E9D0FDBE2D940ECD1C3
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      	sending file:///D:/Data/Desktop/TestBlazorLifeCycle/TestBlazorLifeCycle/Pages/Index.razor 2 A6E1284E44129E9D0FDBE2D940ECD1C3
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      	sending file:///D:/Data/Desktop/TestBlazorLifeCycle/TestBlazorLifeCycle/Pages/Index.razor.cs 2 A6E1284E44129E9D0FDBE2D940ECD1C3
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      	sending file:///D:/Data/Desktop/TestBlazorLifeCycle/TestBlazorLifeCycle/Pages/FetchData.razor 2 A6E1284E44129E9D0FDBE2D940ECD1C3
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      BP req {
        "urlRegex": "[hH][tT][tT][pP][sS]:\\/\\/[lL][oO][cC][aA][lL][hH][oO][sS][tT]:44355\\/\\.\\.\\/[pP][aA][gG][eE][sS]\\/[iI][nN][dD][eE][xX]\\.[rR][aA][zZ][oO][rR]\\.[cC][sS]",
        "lineNumber": 0,
        "columnNumber": 0
      }
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      BP req {
        "urlRegex": "[hH][tT][tT][pP][sS]:\\/\\/[lL][oO][cC][aA][lL][hH][oO][sS][tT]:44355\\/\\.\\.\\/[pP][aA][gG][eE][sS]\\/[iI][nN][dD][eE][xX]\\.[rR][aA][zZ][oO][rR]\\.[cC][sS]",
        "lineNumber": 13,
        "columnNumber": 12
      }
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      BP req {
        "urlRegex": "[fF][iI][lL][eE]:\\/\\/\\/[dD]:\\/[dD][aA][tT][aA]\\/[dD][eE][sS][kK][tT][oO][pP]\\/[tT][eE][sS][tT][bB][lL][aA][zZ][oO][rR][lL][iI][fF][eE][cC][yY][cC][lL][eE]\\/[tT][eE][sS][tT][bB][lL][aA][zZ][oO][rR][lL][iI][fF][eE][cC][yY][cC][lL][eE]\\/[pP][aA][gG][eE][sS]\\/[iI][nN][dD][eE][xX]\\.[rR][aA][zZ][oO][rR]\\.[cC][sS]|[dD]:\\\\[dD][aA][tT][aA]\\\\[dD][eE][sS][kK][tT][oO][pP]\\\\[tT][eE][sS][tT][bB][lL][aA][zZ][oO][rR][lL][iI][fF][eE][cC][yY][cC][lL][eE]\\\\[tT][eE][sS][tT][bB][lL][aA][zZ][oO][rR][lL][iI][fF][eE][cC][yY][cC][lL][eE]\\\\[pP][aA][gG][eE][sS]\\\\[iI][nN][dD][eE][xX]\\.[rR][aA][zZ][oO][rR]\\.[cC][sS]",
        "lineNumber": 13,
        "columnNumber": 12
      }
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      BP req {
        "urlRegex": "[fF][iI][lL][eE]:\\/\\/\\/[dD]:\\/[dD][aA][tT][aA]\\/[dD][eE][sS][kK][tT][oO][pP]\\/[tT][eE][sS][tT][bB][lL][aA][zZ][oO][rR][lL][iI][fF][eE][cC][yY][cC][lL][eE]\\/[tT][eE][sS][tT][bB][lL][aA][zZ][oO][rR][lL][iI][fF][eE][cC][yY][cC][lL][eE]\\/[pP][aA][gG][eE][sS]\\/[iI][nN][dD][eE][xX]\\.[rR][aA][zZ][oO][rR]\\.[cC][sS]|[dD]:\\\\[dD][aA][tT][aA]\\\\[dD][eE][sS][kK][tT][oO][pP]\\\\[tT][eE][sS][tT][bB][lL][aA][zZ][oO][rR][lL][iI][fF][eE][cC][yY][cC][lL][eE]\\\\[tT][eE][sS][tT][bB][lL][aA][zZ][oO][rR][lL][iI][fF][eE][cC][yY][cC][lL][eE]\\\\[pP][aA][gG][eE][sS]\\\\[iI][nN][dD][eE][xX]\\.[rR][aA][zZ][oO][rR]\\.[cC][sS]",
        "lineNumber": 0,
        "columnNumber": 0
      }
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      BP request for 'BreakpointRequest Assembly:  File:  Line: 0 Column: 0' runtime ready True
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      BP request for 'BreakpointRequest Assembly:  File:  Line: 0 Column: 0' runtime ready True
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      BP request for 'BreakpointRequest Assembly:  File:  Line: 0 Column: 0' runtime ready True
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      BP request for 'BreakpointRequest Assembly:  File:  Line: 0 Column: 0' runtime ready True
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      ignoring wasm: Debugger.scriptParsed 
dbug: WebAssembly.Net.Debugging.DevToolsProxy[0]
      DevToolsProxy::Run: Exception System.AggregateException: One or more errors occurred. (The remote party closed the WebSocket connection without completing the close handshake.)
       ---> System.Net.WebSockets.WebSocketException (0x80004005): The remote party closed the WebSocket connection without completing the close handshake.
       ---> System.Net.WebSockets.WebSocketException (0x80004005): The remote party closed the WebSocket connection without completing the close handshake.
         at System.Net.WebSockets.ManagedWebSocket.ThrowIfEOFUnexpected(Boolean throwOnPrematureClosure)
         at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, CancellationToken cancellationToken, Boolean throwOnPrematureClosure)
         at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
         at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
         at WebAssembly.Net.Debugging.DevToolsProxy.ReadOne(WebSocket socket, CancellationToken token)
         --- End of inner exception stack trace ---
         at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
         at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
         at WebAssembly.Net.Debugging.DevToolsProxy.Run(Uri browserUri, WebSocket ideSocket)
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint '/ws-proxy HTTP: GET'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 21589.1053ms 101 
dbug: Microsoft.AspNetCore.Server.Kestrel[10]
      Connection id "0HM0DITMUO9NU" disconnecting.
dbug: Microsoft.AspNetCore.Server.Kestrel[2]
      Connection id "0HM0DITMUO9NU" stopped.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets[7]
      Connection id "0HM0DITMUO9NU" sending FIN because: "The Socket transport's send loop completed gracefully."

```

**Comment by pranavkm** [comment]
@captainsafia could you follow up with the debugging folks?

**Comment by captainsafia** [comment]
Closing this as a dupe of #21849.

This class of issues around debugging external assemblies will likely be resolved once we ship #23121.
